explainer.md

```diff
@@ -131,7 +131,7 @@ For more advanced use cases, such as opening a file in an existing window or dis
 
 ## Permissions API Integration
 
-A `"file-handling"` permission will be defined to protect the user from accidentally allowing a PWA to view files. This permission prompt may asynchronously show right after the user chooses to use a PWA to open a file of any of the file types the PWA tries to associate with, so that the permission context is more understandable and relevant. This permission prompt should also show after the PWA has had an opportunity to load. The prompt should clearly show the PWA's origin, so that the identity of the PWA requesting file read access for the handled file is clear to the user. Launched files should only be queued into the `launchQueue` after the user accepts the permission and allows the PWA to handle the relevant file types. User agents may retain control over default settings and how (or if) they are exposed to the user.
+A `"file-handling"` permission will be defined to protect the user from accidentally allowing a PWA to view files. This permission prompt may asynchronously show right after the user chooses to use a PWA to open a file of any of the file types the PWA tries to associate with, so that the permission context is more understandable and relevant. This permission prompt should also show after the PWA has had an opportunity to load. The prompt should clearly show the PWA's origin, so that the identity of the PWA requesting file read access for the handled file is clear to the user. Launched files should only be queued into the `launchQueue` after the user accepts the permission and allows the PWA to handle the relevant file types. If the user denies the permission, the `launchQueue` consumer should still be notified of the launch but `launchParams.files` will be empty. User agents may retain control over default settings and how (or if) they are exposed to the user.
```

Comment:
I assume we don't want to expose that the user denied the permission.  Is it worth spelling that reasoning out here for purposes of future security review?

Do you think apps will adjust to assume that an empty |files| value will mean that the permission was denied and end up prompting the user to do something about that?  Or, if the permission was denied, would that mean that it just wouldn't be possible to launch the app with the file type handler association (and hence the app would never have its launchQueue consumer called)?

nit: WDYT about making this line wrapped at 80 chars?  I understand that's inconsistent throughout this file but would make future reviews easier.  I'm happy to do that in a separate commit if you're onboard but don't have cycles.

Reply:
> I assume we don't want to expose that the user denied the permission

I think we do want to expose it, and that passing an empty launch queue is the easiest way to do that, since it doesn't require extending the API and there should be no other reason for an empty launch queue. Plus, the example code in the explainer (checking launch queue length and early outting) indicates that perhaps this was the intent all along, but it was never made explicit.

Why I think we want to expose it:

- Based on feedback from partners, PM and UX, it seems we do want apps to know the permission was denied. See discussion on crbug.com/1236075
- The permissions API already exposes it to some extent (the problem here is that it doesn't help the app if the permission denial is not permanent, because it stays on 'prompt' rather than switching to 'denied')
- Other APIs do tend to provide this feedback, I think. Take for example Geolocation with PERMISSION_DENIED: https://developer.mozilla.org/en-US/docs/Web/API/GeolocationPositionError

> Do you think apps will adjust to assume that an empty |files| value will mean that the permission was denied and end up prompting the user to do something about that?

Yes, that is the intent. Personally, I think it makes sense to just close the app window completely in this case (see discussion in bug above). However other parties are in agreement that the app should handle this case with some sort of messaging.

> Or, if the permission was denied, would that mean that it just wouldn't be possible to launch the app with the file type handler association (and hence the app would never have its launchQueue consumer called)?

If the permission is denied permanently, then after that decision the user agent would no longer get this far, i.e. there would not even be an attempt to launch the app to handle the file(s).

> WDYT about making this line wrapped at 80 chars? I understand that's inconsistent throughout this file [...]

What do other WICG explainers do? This file actually does seem consistent in never wrapping. The only newlines are semantic, i.e. at the ends of paragraphs or bullet points.

Reply:
Original wording:
> If the user denies the permission, the `launchQueue` consumer should still be notified of the launch but `launchParams.files` will be an empty array.

How about adding:
> The only case when `launchParams.files` will be empty is when the user has denied the permission or if some unanticipated error has occurred, so an application can use an empty `launchParams.files` array value to infer the user denied the permission.

?  "or if some unanticipated error has occurred" may be unnecessary so feel free to strike but I suppose could happen in the field.

> If the permission is denied permanently, then after that decision the user agent would no longer get this far, i.e. there would not even be an attempt to launch the app to handle the file(s).

This is worth making explicit here if it's behavior that we could spec.

> What do other WICG explainers do? This file actually does seem consistent in never wrapping. The only newlines are semantic, i.e. at the ends of paragraphs or bullet points.

File System Access appears to wrap at X chars: https://raw.githubusercontent.com/WICG/file-system-access/main/EXPLAINER.md

Window Controls Overlay doesn't wrap: https://raw.githubusercontent.com/WICG/window-controls-overlay/main/explainer.md

Local Font Access doesn't wrap: https://raw.githubusercontent.com/WICG/local-font-access/main/README.md

I assume it's a decision left up to explainer authors without guidance from WICG on explainer formatting.  This isn't critical so feel free to punt / ignore for now.

Reply:
Hey just wanted to chime in here. Will this still be work as a signal if other types of launches start using the launch queue? I know WebShareTarget was considering adding support for the `launchQueue` (and we originally designed the API so we could extend it to these other methods).

Maybe it's worth making a note that other launch types should not specify the `files` property, or having some `type` property in the event that consumers can check?

TypeScript below, because my WebIDL is pretty rusty.
```ts
type LaunchEvent = {
    type: 'file',
    files: FileSystemFileHandle[] // or something
} | {
    type: 'share',
    share: WebShareTargetInfo // or something
}
```

I can just imagine a situation where we use an empty files array to indicate the permission was denied but then a new launch method is added, which provides an empty files array (because it doesn't have any files!) and breaks the old permissions handling.

Reply:
> Hey just wanted to chime in here. Will this still work as a signal if other types of launches start using the launch queue? I know WebShareTarget was considering adding support for the `launchQueue` (and we originally designed the API so we could extend it to these other methods).
> 
> Maybe it's worth making a note that other launch types should not specify the `files` property, or having some `type` property in the event that consumers can check?
> 
> TypeScript below, because my WebIDL is pretty rusty.
> 
> ```ts
> type LaunchEvent = {
>     type: 'file',
>     files: FileSystemFileHandle[] // or something
> } | {
>     type: 'share',
>     share: WebShareTargetInfo // or something
> }
> ```
> 
> I can just imagine a situation where we use an empty files array to indicate the permission was denied but then a new launch method is added, which provides an empty files array (because it doesn't have any files!) and breaks the old permissions handling.

Thanks for catching this! All of the above sounds reasonable but I don't think it's strictly necessary. This would only come up if the app specified the same action url for multiple launch handlers. Even if an app wants to use the same page to handle multiple launch types, it can also use that URL to specify which launch type is expected. i.e. the 'action' should be something like '/open_files' or '/open_things?type=textfile'